### ISSUE TITLE
[inductor] [silence] `nn.ConvTranspose2d-F.dropout` outputs inconsistent results with eager

### ISSUE THREAD

**Issue by shaoyuyoung** [pytorch/pytorch#148061]
### 🐛 Describe the bug

**symptom description**: when using `nn.ConvTranspose2d` and `F.dropout` together, outputs are different with eager.
**device backend**: both triton and CPP
**note**: I have used `config.fallback_random = True` and `torch.manual_seed(0)`
**repro**
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config

config.fallback_random = True
torch.set_grad_enabled(False)
torch.manual_seed(0)


class Model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.conv_transpose = torch.nn.ConvTranspose2d(in_channels=3, out_channels=6, kernel_size=3, stride=2, padding=1)

    def forward(self, x):
        torch.manual_seed(0)
        x = self.conv_transpose(x)
        x = F.dropout(x, p=0.5)
        return x


model = Model().cuda()


x = torch.randn(2, 3, 10, 10).cuda()

inputs = [x]


def run_test(model, inputs, backend):
    if backend != "eager":
        model = torch.compile(model, backend=backend)
    torch.manual_seed(0)
    output = model(*inputs)
    return output


output = run_test(model, inputs, 'eager')
c_output = run_test(model, inputs, 'inductor')

print(torch.allclose(output, c_output, 1e-3, 1e-3, equal_nan=True))
print(torch.max(torch.abs(output - c_output)))
```

### Error logs

```
False
tensor(1.7444)
```

### Versions

nightly 20250225

**Comment by desertfire** [comment]
If I add a graph break to the forward function, the test will pass, like,
```
    def forward(self, x):
        x = self.conv_transpose(x)
        torch._dynamo.graph_break()
        x = F.dropout(x, p=0.5)
        return x
```

@eellison , looks like some global state got leaked from conv_transpose to dropout, despite `torch.manual_seed(0)` is called before each run. Do you have any insights?

**Comment by eellison** [comment]
This is due to the conv layout optimization. If you add `torch._inductor.config.layout_optimization = False` this passes.For whatever reason the rng generation of dropout is affected by strides.

The fix is to require the inputs to rng ops as what they were observed in by eager. This is something we should also do for Activation Checkpointing - cc @anijain2305, @zou3519

**Comment by shaoyuyoung** [comment]
`nn.Conv2d` have the same issue:

**repro**
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config

config.fallback_random = True
torch.set_grad_enabled(False)
torch.manual_seed(0)


class Model(nn.Module):

    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(in_channels=3, out_channels=4, kernel_size=3, padding=1)

    def forward(self, x):
        x = self.conv(x)
        x = F.dropout(x, p=0.5)
        return x


model = Model()


x = torch.randn(2, 3, 32, 32)

inputs = [x]


def run_test(model, inputs, device, backend):
    torch.manual_seed(0)
    model = model.to(device)
    inputs = [x.to(device) for x in inputs]
    if backend != "eager":
        model = torch.compile(model, backend=backend)
    torch.manual_seed(0)
    output = model(*inputs)
    return output


device = 'cpu'
output = run_test(model, inputs, device, 'eager')
c_output = run_test(model, inputs, device, 'inductor')
print(torch.allclose(output, c_output, rtol=1e-3, atol=1e-3))
print(torch.max(torch.abs(c_output - output)))

fp64 = run_test(model.to(dtype=torch.float64), [x.to(dtype=torch.float64) for x in inputs], device, 'eager')
print(torch._dynamo.utils.same(output, c_output, fp64))
```
False
tensor(4.0182)
E0419 12:37:56.818000 1607901 site-packages/torch/_dynamo/utils.py:2930] RMSE (res-fp64): 0.83840, (ref-fp64): 0.00000 and shape=torch.Size([2, 4, 32, 32]). res.dtype: torch.float32, multiplier: 2.000000, tol: 0.000100, use_larger_multiplier_for_smaller_tensor: 0
False
```

```

**Comment by shaoyuyoung** [comment]
On nightly20250516, this issue does not occur on triton backend when I use fp64 as the baseline. However, CPP backend seems to output incorrect results still. I have updated to the original repro.
cc @leslie-fang-intel

**Comment by shaoyuyoung** [comment]
No linked PR to this issue...
However, triton backend outputs True when using fp64
```
False
tensor(4.1478, device='cuda:0')
True
```
Additionally, `torch._inductor.config.layout_optimization = False` can eliminate the difference.
Sorry, in this situation, I am not sure whether this is device-specific. :(

**Comment by shaoyuyoung** [comment]
One more word, in **torch 2.6**, triton backend also outputs False in fp64:
```
False
tensor(4.1478, device='cuda:0')
E0516 15:00:27.272000 73749 site-packages/torch/_dynamo/utils.py:2247] RMSE (res-fp64): 0.81181, (ref-fp64): 0.00000 and shape=torch.Size([2, 4, 32, 32]). res.dtype: torch.float32, multiplier: 2.000000, tol: 0.000100, use_larger_multiplier_for_smaller_tensor: 0
False
```
Does this mean this issue is fixed on tirton backend in `nightly202505016`?
However, I don't know which PR fix it. :(

**Comment by jansel** [comment]
My agent was unable to reproduce this on da74fecc24c85f9694061e961858303c44be4338. Reopen if you see this on the latest PyTorch nightly
